internal/api/oidc/key.go

```diff
@@ -19,6 +22,145 @@ import (
+		clock:        clockwork.FromContext(background), // defaults to real clock
+	}
+	go k.purgeOnInterval(background, k.clock.NewTicker(purgeInterval))
+	return k
+}
+
+func (k *keySetCache) purgeOnInterval(background context.Context, ticker clockwork.Ticker) {
+	defer ticker.Stop()
+	for {
+		select {
+		case <-background.Done():
+			return
+		case <-ticker.Chan():
+		}
+
+		// do the actual purging
+		k.mtx.Lock()
+		for instanceID, keys := range k.instanceKeys {
+			for keyID, key := range keys {
+				if key.Expiry().Before(k.clock.Now()) {
+					delete(keys, keyID)
+				}
+			}
+			if len(keys) == 0 {
+				delete(k.instanceKeys, instanceID)
+			}
+		}
+		k.mtx.Unlock()
+	}
+}
+
+func (k *keySetCache) setKey(instanceID, keyID string, key query.PublicKey) {
+	k.mtx.Lock()
+	defer k.mtx.Unlock()
+
+	if keys, ok := k.instanceKeys[instanceID]; ok {
+		keys[keyID] = key
+		return
+	}
+
+	k.instanceKeys[instanceID] = map[string]query.PublicKey{keyID: key}
+}
+
+func (k *keySetCache) getKey(ctx context.Context, keyID string) (_ *jose.JSONWebKey, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.EndWithError(err) }()
+
+	instanceID := authz.GetInstance(ctx).InstanceID()
+
+	k.mtx.RLock()
+	key, ok := k.instanceKeys[instanceID][keyID]
+	k.mtx.RUnlock()
+
+	if ok {
+		if key.Expiry().After(k.clock.Now()) {
+			return jsonWebkey(key), nil
+		}
+		return nil, errors.ThrowInvalidArgument(nil, "OIDC-Zoh9E", "Errors.Key.ExpireBeforeNow")
+	}
+
+	key, err = k.queryKey(ctx, keyID, k.clock.Now())
+	if err != nil {
+		return nil, err
+	}
+	k.setKey(instanceID, keyID, key)
+	return jsonWebkey(key), nil
+}
+
+// VerifySignature implements the oidc.KeySet interface.
+func (k *keySetCache) VerifySignature(ctx context.Context, jws *jose.JSONWebSignature) (_ []byte, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.EndWithError(err) }()
+
+	if len(jws.Signatures) != 1 {
+		return nil, errors.ThrowInvalidArgument(nil, "OIDC-Gid9s", "Errors.Token.Invalid")
+	}
+	key, err := k.getKey(ctx, jws.Signatures[0].Header.KeyID)
+	if err != nil {
+		return nil, err
+	}
+	return jws.Verify(key)
+}
+
+func jsonWebkey(key query.PublicKey) *jose.JSONWebKey {
+	return &jose.JSONWebKey{
+		KeyID:     key.ID(),
+		Algorithm: key.Algorithm(),
+		Use:       key.Use().String(),
+		Key:       key.Key(),
+	}
+}
+
+// keySetMap is a mapping of key IDs to public key data.
+type keySetMap map[string][]byte
+
+// getKey finds the keyID and parses the public key data
+// into a JSONWebKey.
+func (k keySetMap) getKey(keyID string) (*jose.JSONWebKey, error) {
+	pubKey, err := crypto.BytesToPublicKey(k[keyID])
+	if err != nil {
+		return nil, err
+	}
+	return &jose.JSONWebKey{
+		Key:   pubKey,
+		KeyID: keyID,
+		Use:   domain.KeyUsageSigning.String(),
+	}, nil
+}
+
+// VerifySignature implements the oidc.KeySet interface.
+func (k keySetMap) VerifySignature(ctx context.Context, jws *jose.JSONWebSignature) ([]byte, error) {
+	if len(jws.Signatures) != 1 {
```

Comment:
Copied code? Should be use a separate function for both VerifySignature functions?

Reply:
moving a single `if` statement in a seperate function and then check for an error in a new `if` statement doesn't seem like a huge improvement to me?

Reply:
Yes, you're right, I just thought that if you could use a struct definition for both, as using the same errorcode would not make so much of a difference IMO.